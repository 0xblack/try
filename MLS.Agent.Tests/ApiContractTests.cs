using System;
using System.Threading.Tasks;
using Assent;
using WorkspaceServer.Models;
using WorkspaceServer.Models.Execution;
using WorkspaceServer.Tests;
using Xunit;
using Xunit.Abstractions;

namespace MLS.Agent.Tests
{
    public class ApiContractTests : ApiViaHttpTestsBase
    {
        private readonly Configuration configuration;

        public ApiContractTests(ITestOutputHelper output) : base(output)
        {
            configuration = new Configuration()
                            .UsingExtension("json");

#if !DEBUG
            configuration = configuration.SetInteractive(false);
#endif
        }

        [Fact]
        public async Task The_Run_contract_for_compiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode();

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id,
                position: viewport.Position);

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Run_contract_for_noncompiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode("doesn't compile");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id,
                position: viewport.Position);

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        private static Workspace.Buffer EntrypointCode(string mainContent = @"Console.WriteLine(Sample.Method());$$")
        {
            var input = $@"
using System;
using System.Linq;

namespace Example
{{
    public class Program
    {{
        public static void Main()
        {{
            {mainContent}
        }}       
    }}
}}";

            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "EntrypointCode.cs",
                output,
                position ?? 0);
        }

        private static Workspace.Buffer ViewportCode(string methodContent = @"return ""Hello world!"";$$ ")
        {
            var input = $@"
using System.Collections.Generic;
using System;

namespace Example
{{
    public static class Sample
    {{
        public static object Method()
        {{
#region viewport
            {methodContent}
#endregion
        }}
    }}
}}";

            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "ViewportCode.cs",
                output,
                position ?? 0);
        }
    }

}
