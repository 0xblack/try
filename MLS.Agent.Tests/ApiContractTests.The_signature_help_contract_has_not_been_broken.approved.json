{
  "signatures": [
    {
      "name": "Write",
      "label": "void Console.Write(bool value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "bool value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(char value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "char value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(char[] buffer)",
      "documentation": "",
      "parameters": [
        {
          "name": "buffer",
          "label": "char[] buffer",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(char[] buffer, int index, int count)",
      "documentation": "",
      "parameters": [
        {
          "name": "buffer",
          "label": "char[] buffer",
          "documentation": ""
        },
        {
          "name": "index",
          "label": "int index",
          "documentation": ""
        },
        {
          "name": "count",
          "label": "int count",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(decimal value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "decimal value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(double value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "double value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(int value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "int value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(long value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "long value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(object value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "object value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(float value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "float value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "string value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, object arg0)",
      "documentation": "",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg0",
          "label": "object arg0",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, object arg0, object arg1)",
      "documentation": "",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg0",
          "label": "object arg0",
          "documentation": ""
        },
        {
          "name": "arg1",
          "label": "object arg1",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, object arg0, object arg1, object arg2)",
      "documentation": "",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg0",
          "label": "object arg0",
          "documentation": ""
        },
        {
          "name": "arg1",
          "label": "object arg1",
          "documentation": ""
        },
        {
          "name": "arg2",
          "label": "object arg2",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, params object[] arg)",
      "documentation": "",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg",
          "label": "params object[] arg",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(uint value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "uint value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(ulong value)",
      "documentation": "",
      "parameters": [
        {
          "name": "value",
          "label": "ulong value",
          "documentation": ""
        }
      ]
    }
  ],
  "activeSignature": 0,
  "activeParameter": 0
}